hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordMerger.java

```diff
@@ -46,6 +46,28 @@ public interface HoodieRecordMerger extends Serializable {
    */
   Option<Pair<HoodieRecord, Schema>> merge(HoodieRecord older, Schema oldSchema, HoodieRecord newer, Schema newSchema, TypedProperties props) throws IOException;
 
+
+  /**
+   * In some cases a business logic does some checks before flushing a merged record to the disk.
+   * This method does the check and the returned value contains two boolean variables.
+   * <p>
+   * The first variable indicates if the merged record should be flushed to the disk or not.
+   * The second variable takes effect only when the first one is false, and it indicates if
+   * the old record should be kept or not. That is,
+   * (1) (true, _):   the merged one is flushed to the disk; the old record is skipped.
+   * (2) (false, false):  both records skipped, a delete operation.
+   * (3) (false, true):   only the old record flushed to the disk.
+   *
+   * @param record  the merged record.
+   * @param schema  the schema of the merged record.
+   * @return a pair of boolean variables to indicate the flush decision.
+   *
+   * <p> This interface is experimental and might be evolved in the future.
+   **/
+  default Pair<Boolean, Boolean> shouldFlush(HoodieRecord record, Schema schema, TypedProperties props) throws IOException {
```

Comment:
How's this going to evolve over time? What if users want to implement this method and `Pair<Boolean, Boolean>` is not sufficient? Then we add another method?
Also, what's the thinking behind the arguments of this method? If it's going to be simple enough to just decide whether to flush or not, do we need record, schema and props?

Reply:
> If it's going to be simple enough to just decide whether to flush or not

I kind of agree, we can simplify the returned value as a true/false. But maybe @linliu-code has some other considerations here, @linliu-code can you clarify.

> Then we add another method?

I think we may change the method signature directly, by marking this method as expremental, we do not guarantee any compatibility in future versions. And because we have a default impl, so it should be feasible.

Reply:
if we're only going to have 3 cases, can we use an enum instead? I think it can be more clear to the reader and future devs when implementing this method. The overhead of a new `Pair` being created vs the enum singletons can add up as well when processing 10s of thousands of records.

Reply:
> I kind of agree, we can simplify the returned value as a true/false. But maybe @linliu-code has some other considerations here, @linliu-code can you clarify.
@danny0405, @codope, the reason that we need a pair of boolean variables is that:  if a merger decides not to flush the combined record, it faces the question if the old record (the record in the base file) should be kept or not. This question could be very critical, so we should not guess it for the developer who implements their custom merger.

Reply:
> if we're only going to have 3 cases, can we use an enum instead? I think it can be more clear to the reader and future devs when implementing this method. The overhead of a new `Pair` being created vs the enum singletons can add up as well when processing 10s of thousands of records.

@the-other-tim-brown, sounds a good point.

Reply:
> This question could be very critical,

I didn't see such request from any user, even for the contributor from Kuaishou, they just want to keep the merged record or drop it totally. Let's not introduce new semantics if there is no real use case as back-up.

We can evolve the returned value as a `Pair` or `Enum` if there are more feedbacks, at this time point, the behavior for keeping the old record seems not clear to me.

Reply:
> > This question could be very critical,
> 
> I didn't see such request from any user, even for the contributor from Kuaishou, they just want to keep the merged record or drop it totally. Let's not introduce new semantics if there is no real use case as back-up.
> 
> We can evolve the returned value as a `Pair` or `Enum` if there are more feedbacks, at this time point, the behavior for keeping the old record seems not clear to me.

Even in current implementation of `HoodieMergeHandle`, we are still facing this problem: when the `shouldFlush` function returns false,  should we return true or false in `writeRecord` function? Returning true means skipping the old record, false means keeping the old record. No matter which one we choose in advance, we still face the possible situation: what if a user wants the other way?

Reply:
@danny0405 , saw your comment in the slack, we can use the simple signature right now since we can evolve it when we need in the future.